pkg/morph/client/notifications.go

```diff
@@ -1,328 +1,358 @@
-			if !restoreInProgress {
-				c.logger.Info("RPC connection lost, attempting reconnect")
-				if !c.SwitchRPC() {
-					c.logger.Error("can't switch RPC node")
-					break routeloop
-				}
 
-				c.subs.Lock()
-				c.subs.curNotifyChan = make(chan *state.ContainedNotificationEvent)
-				c.subs.curBlockChan = make(chan *block.Block)
-				c.subs.curNotaryChan = make(chan *result.NotaryRequestEvent)
-				go c.restoreSubscriptions(c.subs.curNotifyChan, c.subs.curBlockChan, c.subs.curNotaryChan, restoreCh)
-				c.subs.Unlock()
-				restoreInProgress = true
-			drainloop:
-				for {
-					select {
-					case _, ok := <-notifCh:
-						if !ok {
-							notifCh = nil
-						}
-					case _, ok := <-blCh:
-						if !ok {
-							blCh = nil
-						}
-					case _, ok := <-notaryCh:
-						if !ok {
-							notaryCh = nil
-						}
-					default:
-						break drainloop
-					}
-				}
-			} else { // Avoid getting additional !ok eventc.subs.
-				c.subs.Lock()
-				c.subs.curNotifyChan = nil
-				c.subs.curBlockChan = nil
-				c.subs.curNotaryChan = nil
-				c.subs.Unlock()
-			}
+		if !c.SwitchRPC() {
+			c.logger.Error("can't switch Neo RPC node, stopping event routing")
+			break
 		}
 	}
-	close(c.subs.notifyChan)
-	close(c.subs.blockChan)
-	close(c.subs.notaryChan)
+
+	c.subs.mtx.Lock()
+	closeAllAndClean(c.subs.blocks)
+	closeAllAndClean(c.subs.notifications)
+	closeAllAndClean(c.subs.notaryRequests)
+	c.subs.mtx.Unlock()
 }
 
-// restoreSubscriptions restores subscriptions according to
-// cached information about them.
-func (c *Client) restoreSubscriptions(notifCh chan<- *state.ContainedNotificationEvent,
-	blCh chan<- *block.Block, notaryCh chan<- *result.NotaryRequestEvent, resCh chan<- bool) {
-	var err error
-
-	// new block events restoration
-	if c.subs.subscribedToNewBlocks {
-		_, err = c.client.ReceiveBlocks(nil, blCh)
-		if err != nil {
-			c.logger.Error("could not restore block subscription",
-				zap.Error(err),
-			)
-			resCh <- false
-			return
+func handleEv[Event any, Filter comparable](mtx *sync.RWMutex, e Event, demux eventDemux[Event, Filter], match func(Filter, Event) bool) {
+	mtx.RLock()
+	for kf, subs := range demux {
+		if match(kf, e) {
+			for i := range subs {
+				subs[i].c <- e
+			}
 		}
 	}
+	mtx.RUnlock()
+}
 
-	// notification events restoration
-	for contract := range c.subs.subscribedEvents {
-		_, err = c.client.ReceiveExecutionNotifications(&neorpc.NotificationFilter{Contract: &contract}, notifCh)
-		if err != nil {
-			c.logger.Error("could not restore notification subscription after RPC switch",
-				zap.Error(err),
-			)
-			resCh <- false
-			return
+func closeAllAndClean[Event any, Filter comparable](demux eventDemux[Event, Filter]) {
+	mToClose := make(map[chan<- Event]struct{})
+
+	for f, subs := range demux {
+		for i := range subs {
+			mToClose[subs[i].c] = struct{}{}
 		}
+
+		delete(demux, f)
+	}
+
+	for c := range mToClose {
+		close(c)
 	}
+}
+
+func addSub[Event any, Filter comparable](
```

Comment:
Looks like you're trying to create "stable" subscriptions that survive disconnects again. Please, do not do this, it can't be done reliably. Subscribe to the current client, throw an error and force users to restart upon connection failure. It's _much_ simpler and it's more reliable this way.

Reply:
ok I'll adjust deploy procedure to this flow then

Reply:
done